docs/docs/specs/v0/composer-2.md

```diff
@@ -0,0 +1,389 @@
+  - peer-to-peer, permanent: it is managed by ECHO objects.
+- `AppState` includes current route and params in convenient form.
+- Actions are handled by functions that resemble `reducers` who should return new state reflecting the action committing, and possibly a list of side-effects which may result in more actions or state updates at a later time. (i.e. performing writes to echo objects in a batch is a side-effect, invoking a github API is a side-effect, etc. whereas the indicator that the app is loading something is a purely new state returned from the reducer)
+
+## Plugins, Components, Actions, and other `provides`
+
+- The application framework consumes a list of plugins (statically compiled-in for now).
+- Plugins can provide Components in response to a specific data context: i.e `plugin.provides.component(data: any): React.FC`
+- Compoents provided by plugins can assume they have access to the `AppContext` context and hooks, and may submit actions via `AppContext.dispatch(...actions)`.
+- Plugins can provide action handlers such as `plugin.handlers.[<action-type>](state, action): { state, effects }` and can react to actions coming from other plugins this way.
+- Plugins may provide any number of other data types and APIs which are specific to certain components which may be present in the Application tree at runtime. If a certain component does not exist, then it will not ask for the relevant `provides` from plugins, and they will therefore be inert. (i.e. Plugins provide a `graph` which can be consumed by any number of UI components including the `Tree` which itself is provided by the `TreePlugin` and may represent the entire and total graph built up by all the plugins loaded).
+
+## Example
+
+Below is an example of how the App in the diagram above would be instantiated illustrating how every surface and component comes from a plugin (none are statically defined by the App).
+
+```tsx
+import { RouteComponentProps, useNavigate } from 'react-router';
+
+<Route path="/:spaceKey/:objectId" element={AppContainer} />;
+
+const AppContainer = (props: RouteComponentProps) => {
+  const { spaceKey, objectId } = props.params;
+  const navigate = useNavigate();
+  return (
+    <App
+      plugins={[
+        new SplitViewPlugin(), // provides SplitView with sidebar, main, tools surfaces
+        new TreePlugin(), // provides Tree and Selection which visualize the graph
+        new SpacesPlugin(), // provides graph nodes that represent ECHO spaces and their contents
+        new StackPlugin(), // provides Stack
+        new ComposerPlugin(), // provides Composer
+        new ImagePlugin(), // provides Image
+        new HaloPlugin(), // provides the HALO button and shell in the tools surface
+        new SearchPlugin() // provides the Search element in the tools surface
+      ]}
+      initialState={{
```

Comment:
What's the intention of making this a separate prop, to limit re-renders?

Reply:
in this model the route information is assigned to the state every render via props, and to allow actions to alter the mapping of components-to-surfaces (Say a search term wats to swap tree view for a flat list result view) we can't be resetting/nuking the app state with all the `surfaces:` stuff every render. So thats why its split into `apply this to the state only once to init` and `apply this state to the state every time we render`